listenbrainz_spark/recommendations/candidate_sets.py

```diff
@@ -0,0 +1,384 @@
+
+        Args:
+            artists (tuple): A tuple of artist names.
+            user_id (int): User id of the user.
+
+        Returns:
+            candidate_recording_ids_df (dataframe): Dataframe with columns as:
+                ['user_id', 'recording_id']
+    """
+    df = run_query("""
+        SELECT recording_id
+          FROM recording
+         WHERE artist_name IN %s
+    """ % (artists,))
+    candidate_recording_ids_df = df.withColumn('user_id', lit(user_id)).select('user_id', 'recording_id')
+    return candidate_recording_ids_df
+
+def get_top_artists_recording_ids(similar_artist_df, user_name, user_id):
+    """ Get recording ids of top artists.
+
+        Args:
+            similar_artists_df (dataframe): Dataframe consisting of similar artists.
+            user_name (str): User name of the user.
+            user_id (int): User id of the user.
+
+        Returns:
+            top_artists_recordings_ids_df (dataframe): Dataframe with columns as:
+                ['user_id', 'recording_id']
+
+        Raises:
+            Py4JJavaError if candidate recording ids query cannot be parsed.
+    """
+
+    # top artists with collaborations not equal to zero.
+    top_artists_with_collab = ()
+    top_artist_with_collab_df = similar_artist_df.select('artist_name').distinct()
+    for row in top_artist_with_collab_df.collect():
+        top_artists_with_collab += (row.artist_name,)
+
+    try:
+        if len(top_artists_with_collab) == 1:
+            top_artists_recording_ids_df = get_candidate_recording_ids(tuple((
+                top_artists_with_collab[0])),user_id)
+        else:
+            top_artists_recording_ids_df = get_candidate_recording_ids(top_artists_with_collab, user_id)
+    except Py4JJavaError as err:
+        logging.error('An error occured while fetching recording ids of top artists for "{}". Candidate set \
+            cannot be generated: {} \n {}.'.format(user_name, type(err).__name__, str(err)))
+        return False
+    return top_artists_recording_ids_df
+
+def get_similar_artists_recording_ids(similar_artists_df, top_artists_df, user_name, user_id):
+    """ Get recording ids of similar artists.
+
+        Args:
+            similar_artists_df (dataframe): Dataframe consisting of similar artists.
+            top_artists_df (dataframe) : Dataframe consisting of top artists.
+            user_name (str): User name of the user.
+            user_id (int): User id of the user.
+
+        Returns:
+            similar_artists_recording_ids_df (dataframe): Dataframe with columns as:
+                ['user_id', 'recording_id']
+
+        Raises:
+            Py4JJavaError if candidate recording ids query cannot be parsed.
+    """
+    similar_artists = ()
+    # eliminate artists from similar artists who are a part of top artists
+    similar_artists_df = similar_artists_df.select('similar_artist_name').subtract(top_artists_df.select('artist_name'))
```

Comment:
Instead of using the DataFrame API, we should use SQL here (we prefer SQL over the ORM in all our other projects too)